Firestore/Source/Local/FSTLocalStore.mm

```diff
@@ -177,302 +176,292 @@ - (FSTMaybeDocumentDictionary *)userDidChange:(const User &)user {
-    [localViewReferences removeReferencesToKeys:view.removedKeys forID:targetID];
-  }
-  [self.persistence commitGroup:group];
+  self.persistence.run("NotifyLocalViewChanges", [&]() {
+    FSTReferenceSet *localViewReferences = self.localViewReferences;
+    for (FSTLocalViewChanges *view in viewChanges) {
+      FSTQueryData *queryData = [self.queryCache queryDataForQuery:view.query];
+      FSTAssert(queryData, @"Local view changes contain unallocated query.");
+      FSTTargetID targetID = queryData.targetID;
+      [localViewReferences addReferencesToKeys:view.addedKeys forID:targetID];
+      [localViewReferences removeReferencesToKeys:view.removedKeys forID:targetID];
+    }
+  });
 }
 
 - (nullable FSTMutationBatch *)nextMutationBatchAfterBatchID:(FSTBatchID)batchID {
-  FSTMutationBatch *result = self.persistence.run([&]() -> FSTMutationBatch * {
-    return [self.mutationQueue nextMutationBatchAfterBatchID:batchID];
-  });
+  FSTMutationBatch *result =
+      self.persistence.run("NextMutationBatchAfterBatchID", [&]() -> FSTMutationBatch * {
+        return [self.mutationQueue nextMutationBatchAfterBatchID:batchID];
+      });
   return result;
 }
 
 - (nullable FSTMaybeDocument *)readDocument:(const DocumentKey &)key {
-  FSTWriteGroup *group = [self.persistence startGroupWithAction:@"ReadDocument"];
-  FSTMaybeDocument *result = [self.localDocuments documentForKey:key];
-  [self.persistence commitGroup:group];
-  return result;
+  return self.persistence.run("ReadDocument", [&]() -> FSTMaybeDocument *_Nullable {
+    return [self.localDocuments documentForKey:key];
+  });
 }
 
 - (FSTQueryData *)allocateQuery:(FSTQuery *)query {
-  FSTWriteGroup *group = [self.persistence startGroupWithAction:@"Allocate query"];
-  FSTQueryData *cached = [self.queryCache queryDataForQuery:query];
-  FSTTargetID targetID;
-  FSTListenSequenceNumber sequenceNumber = [self.listenSequence next];
-  if (cached) {
-    // This query has been listened to previously, so reuse the previous targetID.
-    // TODO(mcg): freshen last accessed date?
-    targetID = cached.targetID;
-  } else {
-    targetID = _targetIDGenerator.NextId();
-    cached = [[FSTQueryData alloc] initWithQuery:query
-                                        targetID:targetID
-                            listenSequenceNumber:sequenceNumber
-                                         purpose:FSTQueryPurposeListen];
-    [self.queryCache addQueryData:cached];
-  }
-  [self.persistence commitGroup:group];
+  FSTQueryData *queryData = self.persistence.run("Allocate query", [&]() -> FSTQueryData * {
+    FSTQueryData *cached = [self.queryCache queryDataForQuery:query];
+    FSTTargetID targetID;
+    FSTListenSequenceNumber sequenceNumber = [self.listenSequence next];
+    if (cached) {
+      // This query has been listened to previously, so reuse the previous targetID.
+      // TODO(mcg): freshen last accessed date?
+      targetID = cached.targetID;
```

Comment:
This assignment is now useless, though the TODO is something we should probably still address. Maybe move the TODO into the other branch as something like:

```
TODO(mcg): freshen sequence number on re-listen?

Reply:
Yeah, I actually just noticed this in my next branch. I'll clean this up a bit.